optaplanner-core/src/main/java/org/optaplanner/core/impl/solver/termination/UnimprovedTimeMillisSpentScoreDifferenceThresholdTermination.java

```diff
@@ -86,9 +95,13 @@ public void stepEnded(AbstractStepScope stepScope) {
             for (Iterator<Pair<Long, Score>> it = bestScoreImprovementHistoryQueue.iterator(); it.hasNext(); ) {
                 Pair<Long, Score> bestScoreImprovement = it.next();
                 Score scoreDifference = bestScore.subtract(bestScoreImprovement.getValue());
-                if (scoreDifference.compareTo(unimprovedScoreDifferenceThreshold) >= 0) {
+                final boolean timeLimitNotYetReached =
+                        bestScoreImprovement.getKey() + unimprovedTimeMillisSpentLimit >= bestSolutionTimeMillis;
```

Comment:
Can't this also reuse bestSolutionTimeMillis directly instead of getKey()?

Reply:
Also, optaplanner's code doesn't add "final" for local variables (it's a trade-off between readability (less is more) and avoiding bugs (final to be sure)).
For consistency's sake, I 'd prefer not to do have them on some code, and not on other code, so let's remove the final keywords here.

Reply:
ad `final` : ok
ad reusing `bestSolutionTimeMillis` : there can be no reuse, as `bestSolutionTimeMillis` are taken from the current step, whereas `bestScoreImprovement.getKey()` comes from the last step that improved a score.